### PR TITLE
Preact: Add prop types for Timeago, modify for Selector

### PR DESCRIPTION
**Pull request by caroqliu**
- Adds type definitions for `Timeago` component
- Simplifies `Selector` types by removing the shim types
- Adds a `SelectorDef.ContextProps` 
- Adds prop type for `DateDisplay` component

### CODE REVIEW

File: extensions/amp-date-display/1.0/date-display.js [hunk 1]
```diff
@@ -97,16 +97,10 @@ function getDataForTemplate(props) {
...
-function getEpoch(props) {
-  const {
-    'datetime': datetime = '',
-    'timestampMs': timestampMs = 0,
-    'timestampSeconds': timestampSeconds = 0,
-  } = props;
-
+function getEpoch({datetime = '', timestampMs = 0, timestampSeconds = 0}) {
   let epoch;
   if (datetime.toLowerCase() === 'now') {
```
Comment (caroqliu) [inline extensions/amp-date-display/1.0/date-display.js:105]: This is interesting. The optional type does not complain here. @dvoytenko

Comment (dvoytenko) [reply extensions/amp-date-display/1.0/date-display.js:105]: Object dereferencing allows null/undefined objects. It's been a long-standing Closure issue.

Comment (dvoytenko) [reply extensions/amp-date-display/1.0/date-display.js:105]: Mainly it's function args and return values that are required to fully comply.

File: extensions/amp-selector/1.0/selector.js [hunk 2]
```diff
@@ -94,13 +94,12 @@ export function Option({
...
-    'selected': selected,
-    'selectOption': selectOption,
-    'disabled': selectorDisabled,
-    'multiple': selectorMultiple,
-  } = selectorContext;
+    selected,
+    selectOption,
+    disabled: selectorDisabled,
+    multiple: selectorMultiple,
+  } = /** @type {SelectorDef.ContextProps} */ (useContext(SelectorContext));
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/selector.js:102]: Ready to be fixed.

File: extensions/amp-selector/1.0/selector.type.js [hunk 3]
```diff
@@ -25,6 +25,7 @@ var SelectorDef = {};
  * @typedef {{
  *   as: (string|PreactDef.FunctionalComponent),
  *   disabled: (boolean|undefined),
+ *   domElement: !Element,
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/selector.type.js:28]: Can we add a comment before or after typedef declaration that `domElement` is only used for shimming.

File: extensions/amp-selector/1.0/selector.type.js [hunk 4]
```diff
@@ -34,35 +35,26 @@ var SelectorDef = {};
...
- * }}
- */
-SelectorDef.ShimProps;
-
 /**
  * @typedef {{
  *   as: (string|PreactDef.FunctionalComponent),
  *   option: *,
  *   disabled: (boolean|undefined),
+ *   domElement: !Element,
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/selector.type.js:43]: ditto

File: extensions/amp-selector/1.0/selector.type.js [hunk 5]
```diff
@@ -34,35 +35,26 @@ var SelectorDef = {};
...
  *   onClick: (?function(!Event)|undefined),
  *   role: string,
+ *   selected: (boolean|undefined),
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/selector.type.js:46]: This is also strictly for shimming? I still have some concerns about it. Maybe we should do either `_domElement` and `_selected` to indicate that these are private props. Prefix or suffix ok. Or maybe even more explicitly `shimDomElement` and `shimSelected`?

File: extensions/amp-timeago/1.0/timeago.js [hunk 6]
```diff
@@ -49,25 +53,29 @@ export function Timeago(props) {
...
 /**
  * @param {string} datetime
  * @param {string} locale
- * @param {string} cutoffText
  * @param {number=} opt_cutoff
+ * @param {string=} opt_cutoffText
  * @return {string}
  */
-function getFuzzyTimestampValue(datetime, locale, cutoffText, opt_cutoff) {
+function getFuzzyTimestampValue(datetime, locale, opt_cutoff, opt_cutoffText) {
```
Comment (dvoytenko) [inline extensions/amp-timeago/1.0/timeago.js:70]: These do not seem to be optional (`=`) args - they are always passed from what I can tell. They are just `|undefined` and no need for "opt_" prefix.

**Pull request merged**